src/rmt.rs

```diff
@@ -0,0 +1,415 @@
+
+        use config::Loop;
+        let loop_en = config.looping != Loop::None;
+        let loop_count = match config.looping {
+            Loop::None => 0,
+            Loop::Count(c) => c,
+            Loop::Infinite => 0,
+        };
+
+        let sys_config = rmt_config_t {
+            rmt_mode: rmt_mode_t_RMT_MODE_TX,
+            channel: channel as u32,
+            gpio_num: pin.pin(),
+            clk_div: config.clock_divider,
+            mem_block_num: config.mem_block_num,
+            flags,
+            __bindgen_anon_1: rmt_config_t__bindgen_ty_1 {
+                tx_config: rmt_tx_config_t {
+                    carrier_en,
+                    carrier_freq_hz: carrier.frequency_hz,
+                    carrier_level: carrier.carrier_level as u32,
+                    carrier_duty_percent: carrier.duty_percent,
+                    idle_output_en: config.idle.is_some(),
+                    idle_level: config.idle.map(|i| i as u32).unwrap_or(0),
+                    loop_en,
+                    loop_count,
+                },
+            },
+        };
+
+        unsafe {
+            esp!(rmt_config(&sys_config))?;
+            esp!(rmt_driver_install(channel as u32, 0, 0))?;
+        }
+
+        Ok(Self {
+            channel,
+            items: Default::default(),
+            half_inserted: None,
+        })
+    }
+
+    pub fn counter_clock(&self) -> Result<u32, EspError> {
+        let mut ticks_hz: u32 = 0;
+        esp!(unsafe { rmt_get_counter_clock(self.channel as u32, &mut ticks_hz) })?;
+        Ok(ticks_hz)
+    }
+
+    pub fn add<I>(&mut self, pulses: I) -> Result<(), EspError>
+    where
+        I: IntoIterator<Item = Pulse>,
+    {
+        for pulse in pulses {
+            if let Some(item) = self.half_inserted.as_mut() {
+                // SAFETY: We have retrieved this item which is previously populated with the same
+                // union field.
+                let inner = unsafe { &mut item.__bindgen_anon_1.__bindgen_anon_1 };
+
+                inner.set_level1(pulse.pin_state as u32);
+                inner.set_duration1(pulse.ticks.0 as u32);
+
+                self.flush_half_inserted();
+            } else {
+                let mut inner = rmt_item32_t__bindgen_ty_1__bindgen_ty_1::default();
+                inner.set_level0(pulse.pin_state as u32);
+                inner.set_duration0(pulse.ticks.0 as u32);
+                let item = esp_idf_sys::rmt_item32_t {
+                    __bindgen_anon_1: rmt_item32_t__bindgen_ty_1 {
+                        __bindgen_anon_1: inner,
+                    },
+                };
+                self.half_inserted = Some(item);
+            };
+        }
+
+        Ok(())
+    }
+
+    /// This must be run before starting the RMT pulses otherwise it will drop the odd last entry.
+    fn flush_half_inserted(&mut self) {
+        if let Some(item) = self.half_inserted {
+            self.items.push(item);
```

Comment:
That would be the first driver that requires `alloc` (even though sure, ESP-IDF is of course having `alloc`). 
Also not ideal for performance maybe.

Can we get rid of the `Vec` instance and replace it with e.g. an array? Is there an upper limit on the number of pulses? If not, we can do const generics?

Reply:
> Is there an upper limit on the number of pulses

There's no limit apparently!

> const generics?

So I did a lot of iteration of different ideas, and the most ideal solution I could come up with is to separate the data from the `Writer`, which is a trait. It makes the `Writer` implementation much more succinct.

```rust
pub trait Data {
    fn as_slice(&self) -> &[rmt_item32_t];
}
```

I have two implementations, one use const generics (If I understand correctly, will live on the stack), and one using `Vec` (similar to the previous) for when the user doesn't know the final size.

The neopixel example uses stack based data and morse code example uses the `Vec` variant.

An interesting side note:

I did implement "evaluatable const generics", but ended up crashing the compiler several different ways, so I don't think it's quite stable enough yet! I wanted this implementation it to make it slightly easier for the user to set one `Pulse` at a time instead of a tuple pair. Here's the monstrosity definition of what I'm talking about:

```rust
impl<const N: usize> From<&[Pulse; N]> for StackWriterData<{ (N + 1) / 2 }> {
```